dogstatsd-client/src/lib.rs

```diff
@@ -67,62 +68,57 @@ pub enum DogStatsDAction<'a, T: AsRef<str>, V: IntoIterator<Item = &'a Tag>> {
     Set(T, i64, V),
 }
 
-/// A dogstatsd-client that flushes stats to a given endpoint. Use `new_flusher` to build one.
-#[derive(Debug)]
+/// A dogstatsd-client that flushes stats to a given endpoint.
+#[derive(Debug, Default)]
 pub struct Client {
-    client: StatsdClient,
+    client: Arc<Mutex<Option<StatsdClient>>>,
+    endpoint: Option<Endpoint>,
 }
 
 /// Build a new flusher instance pointed at the provided endpoint.
 /// Returns error if the provided endpoint is not valid.
-pub fn new_flusher(endpoint: Endpoint) -> anyhow::Result<Client> {
+pub fn new(endpoint: Endpoint) -> anyhow::Result<Client> {
+    // defer initialization of the client until the first metric is sent and we definitely know the
+    // client is going to be used to communicate with the endpoint.
     Ok(Client {
-        client: create_client(&endpoint)?,
+        endpoint: Some(endpoint),
+        ..Default::default()
     })
 }
 
 impl Client {
-    /// Set the destination for dogstatsd metrics, if an API Key is provided the client is disabled
-    /// as dogstatsd is not allowed in agentless mode. Returns an error if the provided endpoint
-    /// is invalid.
-    pub fn set_endpoint(&mut self, endpoint: Endpoint) -> anyhow::Result<()> {
```

Comment:
This function wasn't used anywhere else in libdatadog. I can leave it in as it is technically a breaking API change if we think the crate is being used outside of libdatadog.